### DISCUSSION TITLE
CMake Compile with external FMT option

### DISCUSSION THREAD

**Discussion by BobManMasher** [gabime/spdlog#2062] [Q&A] [answered]
Hi,
Compiling latest SPDLOG on an Armel Debian Buster system for use with Gerbera 1.9.0. Gerbera compile warns me I need to use SPDLOG with SPDLOG_FMT_EXTERNAL set to avoid conflicts.  I have compiled the latest FMT (8.0.1) and when I use the variable in my cmake, I get the following message:

'sudo cmake .. -DWITH_SPDLOG_FMT_EXTERNAL=1
-- Build spdlog: 1.9.2
-- Build type: Release
-- Generating example(s)
-- Generating install
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:
    WITH_SPDLOG_FMT_EXTERNAL'

Has the variable been set and will the compile use the external fmt?

**Comment by tt4g** [comment]
`WITH_SPDLOG_FMT_EXTERNAL` is not working.
You need to set CMake variable `SPDLOG_FMT_EXTERNAL` to `ON` for using external fmt library.

**Reply by BobManMasher** [reply]
Thanks for your help,.  I have tried all these ways of setting external FMT:
'-DWITH_SPDLOG_FMT_EXTERNAL=1'
'-DWITH_SPDLOG_FMT_EXTERNAL=ON'
'-DWITH_SPDLOG_FMT_EXTERNAL=YES'

and get the same message i.e.
'CMake Warning:
Manually-specified variables were not used by the project:
WITH_SPDLOG_FMT_EXTERNAL'

When I then compile Gerbera, it also tells me that SPDLOG has been made with its own FMT rather than external.

**Reply by tt4g** [reply]
Replace `-DWITH_SPDLOG_FMT_EXTERNAL=ON` to `-DSPDLOG_FMT_EXTERNAL=ON` (remove `WITH_` prefix).

**Comment by BobManMasher** [comment, marked as answer]
Hi
I run CMAKE as follows:
'cmake -DSPDLOG_FMT_EXTERNAL=ON'

and this is the content of the spdlogConfig.cmake file:

>     cat /usr/local/lib/cmake/spdlog/spdlogConfig.cmake
>     Copyright(c) 2019 spdlog authors
>     Distributed under the MIT License (http://opensource.org/licenses/MIT)
> 
>     ####### Expanded from @PACKAGE_INIT@ by configure_package_config_file() #######
>     ####### Any changes to this file will be overwritten by the next CMake run ####
>     ####### The input file was spdlogConfig.cmake.in ########
> 
>     get_filename_component(PACKAGE_PREFIX_DIR "${CMAKE_CURRENT_LIST_DIR}/../../../" ABSOLUTE)
> 
>     macro(set_and_check _var _file)
>     set(${_var} "${_file}")
>     if(NOT EXISTS "${_file}")
>     message(FATAL_ERROR "File or directory ${_file} referenced by variable ${_var} does not exist !")
>     endif()
>     endmacro()
> 
>     macro(check_required_components _NAME)
>     foreach(comp ${${_NAME}_FIND_COMPONENTS})
>     if(NOT ${NAME}${comp}_FOUND)
>     if(${_NAME}FIND_REQUIRED${comp})
>     set(${_NAME}_FOUND FALSE)
>     endif()
>     endif()
>     endforeach()
>     endmacro()
> 
>     ####################################################################################
> 
>     find_package(Threads REQUIRED)
> 
>     set(SPDLOG_FMT_EXTERNAL OFF)
>     set(config_targets_file spdlogConfigTargets.cmake)
> 
>     if(SPDLOG_FMT_EXTERNAL)
>     include(CMakeFindDependencyMacro)
>     find_dependency(fmt CONFIG)
>     endif()
> 
>     include("${CMAKE_CURRENT_LIST_DIR}/${config_targets_file}")
> 
>     check_required_components(spdlog)
> 

When I compile Gerbera which uses spdlog, I see a message saying that I haven't built spdlog with external fmt.  I am using fmt-8.0.1 BTW which I compilled first and installed.  There seems to be no other fmt packages installed when I do a package check.  Would you please clarify how?

**Reply by tt4g** [reply]
Did you execute `cmake -DSPDLOG_FMT_EXTERNAL=ON` in the directory where you previously built spdlog?
CMake caches some variables, so if you do not download spdlog again and build it in a clean directory, you may get the same artifacts as before.

**Reply by BobManMasher** [reply]
I have a build directory i.e. spdlog-1.9.2/build which I cleared before I started the build last night. The downloaded files are in the directory above so actually my command line is cmake .. -DSPDLOG_FMT_EXTERNAL=ON

**Reply by tt4g** [reply]
Can you find `CMakeCache.txt` in CMake build directory? It records CMake cache variables.
In my case, `SPDLOG_FMT_EXTERNAL:BOOL=ON` is recorded as follows.

```bash
$ mkdir build 
$ cd build
$ cmake .. -DSPDLOG_FMT_EXTERNAL=ON
$ grep -r SPDLOG_FMT_EXTERNAL .
$ grep SPDLOG_FMT_EXTERNAL CMakeCache.txt
SPDLOG_FMT_EXTERNAL:BOOL=ON
SPDLOG_FMT_EXTERNAL_HO:BOOL=OFF
```

**Reply by BobManMasher** [reply]
I have exactly the same on my system:

> root@Nas:/build/spdlog-1.9.2/build# grep SPDLOG_FMT_EXTERNAL CMakeCache.txt
> SPDLOG_FMT_EXTERNAL:BOOL=ON
> SPDLOG_FMT_EXTERNAL_HO:BOOL=OFF

**Reply by tt4g** [reply]
If `SPDLOG_FMT_EXTERNAL=ON` is set only on the command line, then `CMakeCache.txt` and `SPDLOG_FMT_EXTERNAL` in `spdlogConfig.cmake` will have the same value.
Please check the variables of the two files in the build directory and they should have the same value.

My guess is that `/usr/local/lib/cmake/spdlog/spdlogConfig.cmake` is the file generated when you installed spdlog before.
Try to install it again.

**Reply by BobManMasher** [reply]
I don't know what is creating /usr/local/lib/cmake/spdlog/spdlogConfig.cmake but trying all sorts of options and edits to CMakeFiles etc. failed to make an impact. In the end, I purged the spdlog install, deleted the spdlogConfig.cmake file and rebuilt spdlog and was bemused to find spdlogconfig was not recreated.  The build message seems to have gone when I cmake gerbera and the build has now completed and Gerbera 1.9.0 is now running on my test server.
Thanks for your patience and help - Bob
